packages/react-reconciler/src/ReactMutableSource.js

```diff
@@ -0,0 +1,80 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import type {ExpirationTime} from 'react-reconciler/src/ReactFiberExpirationTime';
+import type {FiberRoot} from 'react-reconciler/src/ReactFiberRoot';
+import type {MutableSource, MutableSourceVersion} from 'shared/ReactTypes';
+
+import {isPrimaryRenderer} from './ReactFiberHostConfig';
+import {NoWork} from './ReactFiberExpirationTime';
+
+// Work in progress version numbers only apply to a single render,
+// and should be reset before starting a new render.
+// This tracks which mutable sources need to be reset after a render.
+let workInProgressPrimarySources: Array<MutableSource<any>> = [];
+let workInProgressSecondarySources: Array<MutableSource<any>> = [];
+
+export function clearPendingUpdates(
+  root: FiberRoot,
+  expirationTime: ExpirationTime,
+): void {
+  if (root.mutableSourcePendingUpdateTime <= expirationTime) {
+    root.mutableSourcePendingUpdateTime = NoWork;
+  }
+}
+
+export function getPendingExpirationTime(root: FiberRoot): ExpirationTime {
+  return root.mutableSourcePendingUpdateTime;
+}
+
+export function setPendingExpirationTime(
+  root: FiberRoot,
+  expirationTime: ExpirationTime,
+): void {
+  root.mutableSourcePendingUpdateTime = expirationTime;
```

Comment:
I still feel a little unsure about the suggestion to store all pending updates with a single expiration time.